ofrak_core/ofrak/core/magic.py

```diff
@@ -73,63 +73,102 @@ async def analyze(self, resource: Resource, config=None) -> Magic:
             return Magic(magic_mime, magic_description)
 
 
-class MagicMimeIdentifier(Identifier[None]):
-    """
-    Identify and add the appropriate tag for a given resource based on its mimetype.
+class MagicIdentifier(Identifier[None]):
     """
+    Identify resourced using three identifier patterns:
+
+    1. [MagicMimePattern][ofrak.core.magic.MagicMimePattern]
+    2. [MagicDescriptionPattern][ofrak.core.magic.MagicDescriptionPattern]
+    3. [RawMagicPattern][ofrak.core.magic.RawMagicPattern]
+
+    OFRAK component authors can "register" magic patterns to run whenever this
+    identifier is:
 
-    id = b"MagicMimeIdentifier"
-    targets = (File, GenericBinary)
-    external_dependencies = (LIBMAGIC_DEP,)  # Indirect thru MagicAnalyzer, but worth tagging
+    ```python
+    MagicMimePattern.register(GenericBinary, "application/octet-stream")
+    ```
+    """
 
-    _tags_by_mime: Dict[str, ResourceTag] = dict()
+    targets = (GenericBinary,)
+    external_dependencies = (LIBMAGIC_DEP,)
 
-    async def identify(self, resource: Resource, config=None):
+    async def identify(self, resource: Resource, config=None) -> None:
         _magic = await resource.analyze(Magic)
-        magic_mime = _magic.mime
-        tag = MagicMimeIdentifier._tags_by_mime.get(magic_mime)
+        self._run_magic_mime_patterns(resource, _magic.mime)
+        magic_description = _magic.descriptor
+        self._run_magic_description_pattern(resource, magic_description)
+        await self._run_raw_magic_patterns(resource)
+
+    @staticmethod
+    def _run_magic_mime_patterns(resource: Resource, magic_mime: str):
+        tag = MagicMimePattern.tags_by_mime.get(magic_mime)
         if tag is not None:
             resource.add_tag(tag)
 
+    @staticmethod
+    def _run_magic_description_pattern(resource: Resource, magic_description: str):
+        for matcher, resource_type in MagicDescriptionPattern.matchers.items():
+            if matcher(magic_description):
+                resource.add_tag(resource_type)
+
+    @staticmethod
+    async def _run_raw_magic_patterns(resource: Resource):
+        data_length = min(await resource.get_data_length(), 8)
```

Comment:
Why run the raw magic pattern matchers only against the first 8 bytes? I personally think 32 or 64 would be better.

Should it be documented somewhere that matcher functions will only get the first few bytes, and not all of the data? I would find this surprising as a user, and would expect it to be clearly documented somewhere.